torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -57,11 +103,113 @@ bool isP2POp(OpType opType, bool batchP2P /*= false*/) {
+  TORCH_CHECK(
+      deviceTypeToBackendType_.find(deviceType) !=
+          deviceTypeToBackendType_.end(),
+      "No backend type associated with device type ",
+      deviceType);
+  ProcessGroup::BackendType backendType =
+      deviceTypeToBackendType_.at(deviceType);
+
+  // Check if the backend has already been initialized
+  if (backendTypeToBackend_.find(backendType) != backendTypeToBackend_.end()) {
+    auto backend = backendTypeToBackend_.at(backendType);
+    deviceTypeToBackend_[deviceType] = backend;
+    return backend;
+  } else {
+    // TODO: finish implementing this for lazy initialization
+#ifdef USE_C10D_GLOO
+    std::cout << "C++ initializing the backend for "
+              << backendTypeToStr(backendType) << std::endl;
+    // initialize the backend
+
+    // create a separate prefix store for the newly created backend
+    auto prefixStore = c10::make_intrusive<PrefixStore>(
+        backendTypeToStr(backendType) + "/", store_);
+    c10::intrusive_ptr<Backend> backend;
+
+    // TODO: we should move this into its own function for each respective
+    // backend if backend is gloo then initialize a gloo backend
+    if (backendType == ProcessGroup::BackendType::GLOO) {
+      // create ProcessGroupGloo options
+      backend = ProcessGroupGloo::createProcessGroupGloo(
+          prefixStore, rank_, size_, options_->timeout);
+    }
+
+    // if backend is nccl or gloo then set sequence number
+    if (backendType == ProcessGroup::BackendType::NCCL ||
+        backendType == ProcessGroup::BackendType::GLOO) {
+      backend->setSequenceNumberForGroup();
+    }
+
+    // if using torch_distributed_debug, then wrap backend with
+    // ProcessGroupWrapper
+    if (dist_debug_level_ >= DebugLevel::Detail) {
+      // create new prefix store for the wrapper
+      auto wrapperPrefixStore = c10::make_intrusive<PrefixStore>(
+          backendTypeToStr(backendType) + "/wrapper", store_);
+      auto wrapperBackend = ProcessGroupGloo::createProcessGroupGloo(
+          wrapperPrefixStore, rank_, size_, options_->timeout);
+      backend =
+          c10::make_intrusive<ProcessGroupWrapper>(backend, wrapperBackend);
+    }
+
+    // set internal state and return
+    deviceTypeToBackend_[deviceType] = backend;
+    backendTypeToBackend_[backendType] = backend;
+    return backend;
+
+    // if (backendType == ProcessGroup::BackendType::MPI) {
+    //   auto backend = c10::make_intrusive<ProcessGroupMPI>();
+    // } else if (backendType == ProcessGroup::BackendType::GLOO) {
+    //   return c10::make_intrusive<ProcessGroupGloo>();
+    // } else if (backendType == ProcessGroup::BackendType::NCCL) {
+    //   return c10::make_intrusive<ProcessGroupNCCL>();
+    // } else if (backendType == ProcessGroup::BackendType::UCC) {
+    //   return c10::make_intrusive<ProcessGroupUCC>();
+    // } else {
+    //   TORCH_CHECK(false, "Unknown backend type: ", backendType);
+    // }
+#endif // USE_C10D_GLOO
```

Comment:
Yes to both. 

I only implemented the lazy initialization for backend GLOO (see commit https://github.com/pytorch/pytorch/commit/fbf9feff9e95cf6c6e32d5e4273ad70652ba1303) before hitting the issues that I mentioned (test failures error messages, build failures due to `USE_C10D_GLOO`) and I don't have the backend NCCL implementation but it would be similar.

Given that we are making the feature opt-in, so multiple backends do not need to be created, then we don't need to have lazy initialization for now